v22.1/security-reference/scram-authentication.md

```diff
@@ -0,0 +1,323 @@
+
+In this example, we'll use this [Python script](#appendix-python-scram-hashing-script), which, like many good things, comes from the [open-source community](https://stackoverflow.com/questions/68400120/how-to-generate-scram-sha-256-to-create-postgres-13-user).
+
+### Start a python shell
+
+For example, with docker.
+
+{% include_cached copy-clipboard.html %}
+```bash
+docker run -it python:latest python
+```
+
+### Compute salted password hashes
+
+In the Python shell, use another tool or the [script in the Appendix](#appendix-python-scram-hashing-script) to create a SCRAM-SCHA-256 salted password hash in the format accepted by CockroachDB (which follows PostgreSQL's format).
+
+{% include_cached copy-clipboard.html %}
+```python
+# Either copy-paste the hashing script directly into the shell, or into a file named roach_scram_hasher.py and import by un-commenting the following line
+# from roach_scram_hasher import *
+
+pg_scram_sha256("password")
+```
+```txt
+'SCRAM-SHA-256$119680:D1nWfUQfRP2HWbVuK19Ntw==$ulRLI4fp/bvJN/oBC3x6Q8daLnPit2eWLvm2sex9vyc=:XHVeRmT8iboSn0KMbIx3nMRrebR2gIbi+qEMnHfJi+I='
+```
+
+### Create a SCRAM-SHA-256-authenticated SQL user
+
+In the CockroachDB SQL shell, use the computed hash as the value of `PASSWORD` when [creating a user](../create-user.html) or [rotating the passwors for an existing user](../alter-user.html).
+
+{% include_cached copy-clipboard.html %}
+```sql
+CREATE USER cool_user WITH PASSWORD 'SCRAM-SHA-256$119680:D1nWfUQfRP2HWbVuK19Ntw==$ulRLI4fp/bvJN/oBC3x6Q8daLnPit2eWLvm2sex9vyc=:XHVeRmT8iboSn0KMbIx3nMRrebR2gIbi+qEMnHfJi+I=';
+
+ALTER USER other_user WITH PASSWORD "$scram$100000$vZcyRiil1BojRGhN$sha-1=pRLECKY8rLwZKkmhKu02b8RfIeQ,sha-256=ClkmygvxmE9tOx8swe2lD5kb/4rjE3qa2BF6mAhshHE,sha-512=l3zLEbJLHqRZGi5uCdwDqkNzk4bfoZTcvyKRXyu6YP3mtoh7JFclxKdAKXKQTBAWMQDv.pbNgUt3/KOX4Hgteg";
+```
+```
+CREATE ROLE
+
+Time: 156ms total (execution 156ms / network 0ms)
+```
+
+### Inspect the user
+
+In this way, the server stores the information needed for SCRAM authentication, despite the plaintext password never having been conveyed to CockroachDB. The 'concern' with plaintext passwords has been cleanly 'separated' from CockroachDB.
+
+{% include_cached copy-clipboard.html %}
+```sql
+SELECT username, "hashedPassword" FROM system.users WHERE username='cool_user'; 
+SELECT *  FROM system.users WHERE username = "cool_user";
+```
+```
+  username  |                                                             hashedPassword
+------------+------------------------------------------------------------------------------------------------------------------------------------------
+  cool_user | SCRAM-SHA-256$119680:6XN0y7A83hylmX+3uMRPvQ==$mOI18rsucBjSrNMDDqNJ1/q4cROvRDLIUqjC0BzQHEg=:4zPCVFflWWHD53ZEIrcH4q3X1+le86TBIV0Dce6fIuc=
+```
+
+## Appendix: Python SCRAM-hashing script
```

Comment:
Do we use appendices elsewhere? I was wondering if this is necessary and you couldn't just plop it where you mention it.... but I also see how it could be a little distracting there. Might be worth opening a docs style issue as a "Should we consider using code appendix when it is not necessarily part of the workflow but helpful for users". Is there a source for this code on GitHub?

Reply:
But that's non-blocking, just thought it was interesting!